lighthouse-core/gather/computed/dependency-graph/estimator/tcp-connection.js

```diff
@@ -0,0 +1,147 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const INITIAL_CONGESTION_WINDOW = 10;
+const TCP_SEGMENT_SIZE = 1460;
+
+class TcpConnection {
+  /**
+   * @param {number} rtt
+   * @param {number} throughput
+   * @param {number=} responseTime
+   * @param {boolean=} ssl
+   */
+  constructor(rtt, throughput, responseTime = 0, ssl = true) {
+    this._warmed = false;
+    this._ssl = ssl;
+    this._rtt = rtt;
+    this._availableThroughput = throughput;
+    this._responseTime = responseTime;
+    this._congestionWindow = INITIAL_CONGESTION_WINDOW;
+  }
+
+  /**
+   * @param {number} rtt
+   * @param {number} availableThroughput
+   * @return {number}
+   */
+  static maximumSaturatedConnections(rtt, availableThroughput) {
+    const roundTripsPerSecond = 1000 / rtt;
+    const bytesPerRoundTrip = TCP_SEGMENT_SIZE;
+    const bytesPerSecond = roundTripsPerSecond * bytesPerRoundTrip;
+    const minimumThroughputRequiredPerRequest = bytesPerSecond * 8;
+    return Math.floor(availableThroughput / minimumThroughputRequiredPerRequest);
+  }
+
+  /**
+   * @return {number}
+   */
+  _computeMaximumCongestionWindowInSegments() {
+    const bytesPerSecond = this._availableThroughput / 8;
+    const secondsPerRoundTrip = this._rtt / 1000;
+    const bytesPerRoundTrip = bytesPerSecond * secondsPerRoundTrip;
+    return Math.floor(bytesPerRoundTrip / TCP_SEGMENT_SIZE);
+  }
+
+  /**
+   * @param {number} throughput
+   */
+  setThroughput(throughput) {
+    this._availableThroughput = throughput;
+  }
+
+  /**
+   * @param {number} congestion
+   */
+  setCongestionWindow(congestion) {
+    this._congestionWindow = congestion;
+  }
+
+  /**
+   * @param {boolean} warmed
+   */
+  setWarmed(warmed) {
+    this._warmed = warmed;
+  }
+
+  /**
+   * Simulates a network download of a particular number of bytes over an optional maximum amount of time
+   * and returns information about the ending state.
+   *
+   * See https://hpbn.co/building-blocks-of-tcp/#three-way-handshake and
+   *  https://hpbn.co/transport-layer-security-tls/#tls-handshake for details.
+   *
+   * @param {number} bytesToDownload
+   * @param {number=} timeAlreadyElapsed
+   * @param {number=} maximumTimeToElapse
+   * @return {{timeElapsed: number, roundTrips: number, bytesDownloaded: number, congestionWindow: number}}
+   */
+  calculateTimeToDownload(bytesToDownload, timeAlreadyElapsed = 0, maximumTimeToElapse = Infinity) {
+    const twoWayLatency = this._rtt;
+    const oneWayLatency = twoWayLatency / 2;
+    const maximumCongestionWindow = this._computeMaximumCongestionWindowInSegments();
+
+    let handshakeAndRequest = oneWayLatency;
+    if (!this._warmed) {
+      handshakeAndRequest =
```

Comment:
I'm not a huge fan of splitting it since the handshake is in fact concluding in the same trip as the request is being made, adding a comment with `// technically the ACK is still happening blah blah` seems a tad more to process than a variable name indicating the two are conjoined IMO. It also slightly complicates the later http2 case where neither are required once the connection has been warmed.

Do you think addressing the other TTFB issues makes this clearer or you'd still like to see the split?

Reply:
agreed to keep `handshakeAndRequest` as a unit as it's more accurate to the reality of TCP. 👍